go/vt/vtadmin/api.go

```diff
@@ -221,13 +221,42 @@ func (api *API) GetKeyspaces(ctx context.Context, req *vtadminpb.GetKeyspacesReq
 				return
 			}
 
-			m.Lock()
+			kss := make([]*vtadminpb.Keyspace, 0, len(resp.Keyspaces))
+
+			var (
+				kwg sync.WaitGroup
+				km  sync.Mutex
+			)
+
 			for _, ks := range resp.Keyspaces {
-				keyspaces = append(keyspaces, &vtadminpb.Keyspace{
-					Cluster:  c.ToProto(),
-					Keyspace: ks,
-				})
+				kwg.Add(1)
+
+				// Find all shards for each keyspace in the cluster, in parallel
+				go func(c *cluster.Cluster, ks *vtctldatapb.Keyspace) {
+					defer kwg.Done()
+					sr, err := c.Vtctld.FindAllShardsInKeyspace(ctx, &vtctldatapb.FindAllShardsInKeyspaceRequest{
+						Keyspace: ks.Name,
+					})
+
+					if err != nil {
+						er.RecordError(err)
```

Comment:
Is it ok to use the same `concurrency.AllErrorRecorder` in nested goroutines? 🤔

Reply:
Yep! It's [completely synchronized](https://github.com/vitessio/vitess/blob/master/go/vt/concurrency/error_recorder.go#L88-L90)